src/service-worker.js

```diff
@@ -1,33 +1,61 @@
 /* eslint-env serviceworker */
+self.PusherPushNotifications = {
+  onNotificationReceived: null,
+};
 
-self.addEventListener('push', function(e) {
-  const payload = e.data.json();
-
-  const title = payload.notification.title || '';
-  const body = payload.notification.body || '';
-  const icon = payload.notification.icon;
-  const data = payload.data || {};
+self.addEventListener('push', e => {
+  let payload;
+  try {
+    payload = e.data.json();
+  } catch (_) {
+    return; // Not a pusher notification
+  }
 
-  if (payload.notification.deep_link) {
-    // Copying the deep_link into the data payload so that it can
-    // be accessed in the notificationclick handler.
-    data.pusher.deep_link = payload.notification.deep_link;
+  if (!payload.data || !payload.data.pusher) {
+    return; // Not a pusher notification
   }
 
-  const options = {
-    title,
-    body,
-    icon,
-    data,
+  const customerPayload = { ...payload };
+  const customerData = {};
+  Object.keys(customerPayload.data || {}).forEach(key => {
+    if (key !== 'pusher') {
+      customerData[key] = customerPayload.data[key];
+    }
+  });
+  customerPayload.data = customerData;
+
+  const handleNotification = payload => {
+    const title = payload.notification.title || '';
+    const body = payload.notification.body || '';
+    const icon = payload.notification.icon;
+
+    const options = {
+      body,
+      icon,
+      data: { pusherPayload: payload },
+    };
+
+    e.waitUntil(self.registration.showNotification(title, options));
```

Comment:
Would we want to call `self.PusherPushNotifications.onNotificationReceived` in a `waitUntil`?

Reply:
Or perhaps give the user some way of doing that in their callback?

Reply:
That is a good point. Maybe we just give them the whole event? Seems the most flexible